src/vs/editor/contrib/linesOperations/linesOperations.ts

```diff
@@ -672,146 +679,132 @@ export class JoinLinesAction extends EditorAction {
-
-					trimmedLinesContent += (insertSpace ? ' ' : '') + lineTextWithoutIndent;
-
-					if (insertSpace) {
-						columnDeltaOffset = lineTextWithoutIndent.length + 1;
-					} else {
-						columnDeltaOffset = lineTextWithoutIndent.length;
-					}
-				} else {
-					columnDeltaOffset = 0;
 				}
+				if (removeLines[i].endLineNumber <= model.getLineCount()) {
+					string.push(model.getLineContent(removeLines[i].endLineNumber).trim());
+				}
+				conString.push(string.join(' ') + ' ');
 			}
 
-			let deleteSelection = new Range(startLineNumber, startColumn, endLineNumber, endColumn);
+		}
 
-			if (!deleteSelection.isEmpty()) {
-				let resultSelection: Selection;
+		if (ops.length === 0) {
+			return;
+		}
 
-				if (selection.isEmpty()) {
-					edits.push(EditOperation.replace(deleteSelection, trimmedLinesContent));
-					resultSelection = new Selection(deleteSelection.startLineNumber - lineOffset, trimmedLinesContent.length - columnDeltaOffset + 1, startLineNumber - lineOffset, trimmedLinesContent.length - columnDeltaOffset + 1);
-				} else {
-					if (selection.startLineNumber === selection.endLineNumber) {
-						edits.push(EditOperation.replace(deleteSelection, trimmedLinesContent));
-						resultSelection = new Selection(selection.startLineNumber - lineOffset, selection.startColumn,
-							selection.endLineNumber - lineOffset, selection.endColumn);
-					} else {
-						edits.push(EditOperation.replace(deleteSelection, trimmedLinesContent));
-						resultSelection = new Selection(selection.startLineNumber - lineOffset, selection.startColumn,
-							selection.startLineNumber - lineOffset, trimmedLinesContent.length - selectionEndPositionOffset);
-					}
-				}
+		let linesDeleted = 0;
+		for (let i = 0; i < removeLines.length; i++) {
+			const op = removeLines[i];
 
-				if (Range.intersectRanges(deleteSelection, primaryCursor) !== null) {
-					endPrimaryCursor = resultSelection;
-				} else {
-					endCursorState.push(resultSelection);
-				}
+			let startLineNumber = op.startLineNumber;
+			let endLineNumber = op.endLineNumber;
+
+			if (endLineNumber === startLineNumber && endLineNumber < model.getLineCount()) {
+				endLineNumber += 1;
 			}
 
-			lineOffset += deleteSelection.endLineNumber - deleteSelection.startLineNumber;
+			let start = model.getLineFirstNonWhitespaceColumn(startLineNumber);
+			if (endLineNumber < model.getLineCount()) {
+				cursorState.push(new Selection(startLineNumber - linesDeleted, start, startLineNumber - linesDeleted, start + conString[i].length));
+				linesDeleted += (endLineNumber - startLineNumber);
+			} else {
+				cursorState.push(new Selection(startLineNumber - linesDeleted, start, startLineNumber - linesDeleted, start + conString[conString.length - 1].length));
+			}
 		}
 
-		endCursorState.unshift(endPrimaryCursor);
 		editor.pushUndoStop();
-		editor.executeEdits(this.id, edits, endCursorState);
+		editor.executeEdits(this.id, ops, cursorState);
```

Comment:
This function is now rather large.  Can be be broken down into smaller private functions?